# **Initial Data Cleaning**

## **Setting Up Environment**


### 1. Imprting Essential liberaries

In [ ]:
import numpy as np  
import pandas as pd  
import os  
import time  
import re 

### 2. Reading the original CSV files

In [51]:
df_customer = pd.read_csv(r"C:\Mine\My Projects\E-commerce database\CSV files\customer_dim.csv")
df_fact = pd.read_csv(r"C:\Mine\My Projects\E-commerce database\CSV files\fact_table.csv")
df_item = pd.read_csv(r"C:\Mine\My Projects\E-commerce database\CSV files\item_dim.csv")
df_store = pd.read_csv(r"C:\Mine\My Projects\E-commerce database\CSV files\store_dim.csv")
df_time = pd.read_csv(r"C:\Mine\My Projects\E-commerce database\CSV files\time_dim.csv")
df_trans = pd.read_csv(r"C:\Mine\My Projects\E-commerce database\CSV files\trans_dim.csv")


## **Data Inspection and Cleaning**

### **1. Customer table**

In [52]:
df_customer.head()

customer_key             name     contact_no            nid
0      C000001            sumit  8801920345851  7505075708899
1      C000002         tammanne  8801817069329  1977731324842
2      C000003    kailash kumar  8801663795774  3769494056318
3      C000004  bhagwati prasad  8801533627961  9378834712725
4      C000005             ajay  8801943715786  3540815556323

In [53]:
df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9191 entries, 0 to 9190
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   customer_key  9191 non-null   object
 1   name          9164 non-null   object
 2   contact_no    9191 non-null   int64 
 3   nid           9191 non-null   int64 
dtypes: int64(2), object(2)
memory usage: 287.3+ KB


In [54]:
if df_customer['customer_key'].duplicated().any():
    print("Duplicate customer keys found!")
else :
    print("No Duplicated Found")

No Duplicated Found


In [55]:
df_customer.name.isnull().sum()

27

In [56]:
df_customer['name'].fillna('Anonymous Customer', inplace=True)

C:\Users\ahmed\AppData\Local\Temp\ipykernel_11640\1986894991.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_customer['name'].fillna('Anonymous Customer', inplace=True)


In [57]:
df_customer.name.isnull().sum()

0

In [58]:
def manipulate_strings(column):
    column_cleaned = re.sub(r'[^a-zA-Z\s]', '', column)
    case = column_cleaned.title()
    return case

In [59]:
df_customer['name'] = df_customer['name'].apply(manipulate_strings)
df_customer['name']

0                 Sumit
1              Tammanne
2         Kailash Kumar
3       Bhagwati Prasad
4                  Ajay
             ...       
9186          Smt Radha
9187       Tulsi Sharma
9188          Madhubala
9189         Pintu Devi
9190              Roopa
Name: name, Length: 9191, dtype: object

In [60]:
df_customer['contact_no'] = df_customer['contact_no'].astype(str)
df_customer['nid'] = df_customer['nid'].astype(str)

## **Summary**

| Category             | State | Action |
|----------------------|-------|--------|
| Number of duplicates | None  | None   |
| Null values          | 27 (name) | Filled with "Anonymous Customer" |
| Names values inspection | Need reformatting | Applied a defined manipulate_strings formula |

---------------------

### **2. Fact Table**

In [61]:
df_fact.head()

payment_key customer_key time_key item_key store_key  quantity   unit  \
0        P026      C004510  T049189   I00177    S00307         1     ct   
1        P022      C008967  T041209   I00248    S00595         1  rolls   
2        P030      C007261   T03633   I00195    S00496         8     ct   
3        P032      C007048  T084631   I00131     S0086         8     ct   
4        P014      C006430  T071276   I00050    S00488         8   cans   

   unit_price  total_price  
0        35.0         35.0  
1        26.0         26.0  
2        12.5        100.0  
3        14.0        112.0  
4         8.0         64.0

In [62]:
df_fact.describe()

quantity      unit_price     total_price
count  1000000.000000  1000000.000000  1000000.000000
mean         6.000185       17.561602      105.401436
std          3.161932        8.637097       80.829301
min          1.000000        6.000000        6.000000
25%          3.000000       13.000000       47.250000
50%          6.000000       16.000000       90.000000
75%          9.000000       20.000000      144.000000
max         11.000000       55.000000      605.000000

In [63]:
df_fact.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 9 columns):
 #   Column        Non-Null Count    Dtype  
---  ------        --------------    -----  
 0   payment_key   1000000 non-null  object 
 1   customer_key  1000000 non-null  object 
 2   time_key      1000000 non-null  object 
 3   item_key      1000000 non-null  object 
 4   store_key     1000000 non-null  object 
 5   quantity      1000000 non-null  int64  
 6   unit          996277 non-null   object 
 7   unit_price    1000000 non-null  float64
 8   total_price   1000000 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 68.7+ MB


In [64]:
df_fact.sample(5)

payment_key customer_key time_key item_key store_key  quantity  \
896816        P004      C001611  T067681   I00174    S00170         4   
255379        P013      C006190   T00334   I00065    S00568         3   
348406        P034      C006709  T043938   I00174    S00302        11   
741079        P019      C008202  T053098   I00245    S00471         6   
709105        P004      C003017  T016356   I00169    S00617         2   

           unit  unit_price  total_price  
896816       oz        18.0         72.0  
255379  bottles        20.0         60.0  
348406       oz        18.0        198.0  
741079    rolls        26.0        156.0  
709105       ct        17.0         34.0

In [65]:
df_fact['item_key'].value_counts()

item_key
I00015    3952
I00203    3939
I00082    3923
I00257    3920
I00230    3913
          ... 
I00168    3666
I00141    3665
I00068    3665
I00173    3659
I00051    3625
Name: count, Length: 264, dtype: int64

In [66]:
df_fact['unit'].value_counts()

unit
ct          355938
cans        215602
bottles     140073
oz           71956
bags         49739
pk           37918
ct.          30474
bars         18950
pack         18876
Bags         11298
tubs         11271
rolls         7624
Ct            7575
cartons       3893
botlltes      3796
lb            3795
tins          3772
oz.           3727
Name: count, dtype: int64

In [67]:
df_fact.groupby('store_key')['total_price'].sum().sort_values()

store_key
S00182    130082.75
S00566    131603.50
S00550    131824.00
S00100    132461.50
S00310    133580.50
            ...    
S0080     157377.50
S0039     157483.00
S00328    157692.00
S00601    158027.50
S0010     159409.00
Name: total_price, Length: 726, dtype: float64

In [68]:
df_fact[df_fact.isnull().any(axis=1)]

payment_key customer_key time_key item_key store_key  quantity unit  \
413           P030      C001653  T053668   I00158    S00327         1  NaN   
478           P013      C008861  T036513   I00158    S00280         9  NaN   
1500          P007      C008864  T052262   I00158    S00308         7  NaN   
2033          P037      C002870  T050819   I00158    S00506         7  NaN   
2053          P020      C003666  T094595   I00158     S0077        11  NaN   
...            ...          ...      ...      ...       ...       ...  ...   
998455        P016      C007003   T04678   I00158    S00522         1  NaN   
998754        P039      C004215  T022792   I00158     S0092         9  NaN   
998757        P027      C000309   T07076   I00158    S00550         9  NaN   
999136        P033      C008559  T070336   I00158    S00553        11  NaN   
999660        P019      C002022   T04150   I00158    S00347         7  NaN   

        unit_price  total_price  
413           17.0         17.0  
478           17.0        153.0  
1500          17.0        119.0  
2033          17.0        119.0  
2053          17.0        187.0  
...            ...          ...  
998455        17.0         17.0  
998754        17.0        153.0  
998757        17.0        153.0  
999136        17.0        187.0  
999660        17.0        119.0  

[3723 rows x 9 columns]

In [69]:
df_fact['unit'].fillna("pack", inplace=True)

C:\Users\ahmed\AppData\Local\Temp\ipykernel_11640\3921156335.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_fact['unit'].fillna("pack", inplace=True)


In [70]:
df_fact[df_fact['item_key']=='I00158']

payment_key customer_key time_key item_key store_key  quantity  unit  \
413           P030      C001653  T053668   I00158    S00327         1  pack   
478           P013      C008861  T036513   I00158    S00280         9  pack   
1500          P007      C008864  T052262   I00158    S00308         7  pack   
2033          P037      C002870  T050819   I00158    S00506         7  pack   
2053          P020      C003666  T094595   I00158     S0077        11  pack   
...            ...          ...      ...      ...       ...       ...   ...   
998455        P016      C007003   T04678   I00158    S00522         1  pack   
998754        P039      C004215  T022792   I00158     S0092         9  pack   
998757        P027      C000309   T07076   I00158    S00550         9  pack   
999136        P033      C008559  T070336   I00158    S00553        11  pack   
999660        P019      C002022   T04150   I00158    S00347         7  pack   

        unit_price  total_price  
413           17.0         17.0  
478           17.0        153.0  
1500          17.0        119.0  
2033          17.0        119.0  
2053          17.0        187.0  
...            ...          ...  
998455        17.0         17.0  
998754        17.0        153.0  
998757        17.0        153.0  
999136        17.0        187.0  
999660        17.0        119.0  

[3723 rows x 9 columns]

In [71]:
df_fact.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 9 columns):
 #   Column        Non-Null Count    Dtype  
---  ------        --------------    -----  
 0   payment_key   1000000 non-null  object 
 1   customer_key  1000000 non-null  object 
 2   time_key      1000000 non-null  object 
 3   item_key      1000000 non-null  object 
 4   store_key     1000000 non-null  object 
 5   quantity      1000000 non-null  int64  
 6   unit          1000000 non-null  object 
 7   unit_price    1000000 non-null  float64
 8   total_price   1000000 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 68.7+ MB


## **Summary**

| Category             | State | Action |
|----------------------|-------|--------|
| Number of duplicates | None  | None   |
| Null values          | 3723 (unit) | Filled with `pack` Look beneath the table* |

After deep investigation, we found that all the null values in the units column carry the same `item_key`: `I00158` and it costs `$17.0`, so we looked up the `item table` to find that it's also null.

This item's name is `Frito Bold Flavors Variety`, so we googled it to find that it's sold as `pack`.

---------------------------

# **Items**

In [72]:
df_item.sample(10)

item_key                               item_name          description  \
78    I00079        Pure Leaf Unsweetened Tea 18.5oz  Beverage - Iced Tea   
72    I00073  Gatorade X-Factor Fierce Variety 20 oz  Beverage - Gatorade   
106   I00107            Dunkin Donuts Original Blend        Coffee Ground   
247   I00248               Charmin Ultra Bath Tissue     Kitchen Supplies   
108   I00109            Folgers Classic Roast Coffee        Coffee Ground   
94    I00095               Propel Zero Variety 20 oz       Beverage Water   
239   I00240           Ice Breaker Mints Sugar Free           Gum - Mints   
205   I00206   Nature Valley Soft Oatmeal Squares          Food - Healthy   
184   I00185  Apple Sauce Variety GoGO Squeeze 3.2oz       Food - Healthy   
17    I00018            Fresca Original - 12 oz cans   a. Beverage - Soda   

     unit_price    man_country                        supplier      unit  
78        17.00          China            CHERRY GROUP CO.,LTD   bottles  
72        20.00       Cambodia  NINGBO SEDUNO IMP & EXP CO.LTD   bottles  
106       20.00         poland              CHROMADURLIN S.A.S        oz  
247       26.00     Bangladesh                    DENIMACH LTD     rolls  
108       17.00        Germany              Friedola 1888 GmbH        oz  
94        16.00        Finland                     HARDFORD AB  botlltes  
239       14.00  United States                       MAESA SAS        ct  
205       15.00        Germany              Friedola 1888 GmbH        ct  
184       14.00          India       Indo Count Industries Ltd        ct  
17         6.75          India       Indo Count Industries Ltd      cans

In [73]:
df_item.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 264 entries, 0 to 263
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   item_key     264 non-null    object 
 1   item_name    264 non-null    object 
 2   description  264 non-null    object 
 3   unit_price   264 non-null    float64
 4   man_country  264 non-null    object 
 5   supplier     264 non-null    object 
 6   unit         263 non-null    object 
dtypes: float64(1), object(6)
memory usage: 14.6+ KB


In [74]:
df_item[df_item.isnull().any(axis=1)]

item_key                   item_name   description  unit_price  \
157   I00158  Frito Bold Flavors Variety  Food - Chips        17.0   

    man_country  supplier unit  
157   Lithuania  BIGSO AB  NaN

In [75]:
df_item['unit'].fillna("pack", inplace=True)


C:\Users\ahmed\AppData\Local\Temp\ipykernel_11640\951068782.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_item['unit'].fillna("pack", inplace=True)


In [76]:
df_item['man_country'].unique()

array(['Netherlands', 'poland', 'Bangladesh', 'Finland', 'Lithuania',
       'India', 'Germany', 'China', 'United States', 'Cambodia'],
      dtype=object)

In [77]:
df_item_sorted = df_item.sort_values(by='unit_price', ascending=False)


In [78]:
df_item_sorted.head()

item_key                              item_name  \
60    I00061                          Red Bull 12oz   
114   I00115    K Cups Daily Chef Columbian Supremo   
118   I00119  K Cups Original Donut Shop Med. Roast   
115   I00116      K Cups Dunkin Donuts Medium Roast   
116   I00117        K Cups Folgers Lively Columbian   

                   description  unit_price    man_country  \
60   Beverage - Energy/Protein        55.0  United States   
114              Coffee K-Cups        53.0          India   
118              Coffee K-Cups        53.0          India   
115              Coffee K-Cups        48.0  United States   
116              Coffee K-Cups        46.0       Cambodia   

                           supplier  unit  
60                        MAESA SAS  cans  
114       Indo Count Industries Ltd    ct  
118       Indo Count Industries Ltd    ct  
115                       MAESA SAS    ct  
116  NINGBO SEDUNO IMP & EXP CO.LTD    ct

In [79]:
df_count = df_item['man_country'].value_counts()
df_count

man_country
Bangladesh       34
India            32
Germany          31
poland           28
Finland          27
Netherlands      26
Lithuania        26
Cambodia         24
United States    20
China            16
Name: count, dtype: int64

| Category             | State | Action |
|----------------------|-------|--------|
| Number of duplicates | None  | None   |
| Null values          | 1 (unit) | Filled with `pack` after looking up the item name|

-----

# **Stores**

In [80]:
df_store.head()

store_key division  district         upazila
0     S0001   SYLHET  HABIGANJ      AJMIRIGANJ
1     S0002   SYLHET  HABIGANJ         BAHUBAL
2     S0003   SYLHET  HABIGANJ      BANIACHONG
3     S0004   SYLHET  HABIGANJ     CHUNARUGHAT
4     S0005   SYLHET  HABIGANJ  HABIGANJ SADAR

In [81]:
df_store.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 726 entries, 0 to 725
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   store_key  726 non-null    object
 1   division   726 non-null    object
 2   district   726 non-null    object
 3   upazila    726 non-null    object
dtypes: object(4)
memory usage: 22.8+ KB


In [82]:
df_store['store_key'] = df_store['store_key'].astype("str")

## **Summary**

>**This table looked pretty clean**

-----------------------

# **Time**

In [83]:
df_time.head()

time_key                 date  hour  day  week  month  quarter  year
0   T00001  2017-05-20 14:56:00    14   20     3      5        2  2017
1   T00002  2015-01-30 22:14:00    22   30     4      1        1  2015
2   T00003  2020-03-14 02:34:00     2   14     2      3        1  2020
3   T00004  2018-04-27 12:19:00    12   27     4      4        2  2018
4   T00005  2018-04-14 10:43:00    10   14     2      4        2  2018

In [84]:
df_time.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   time_key  99999 non-null  object
 1   date      99999 non-null  object
 2   hour      99999 non-null  int64 
 3   day       99999 non-null  int64 
 4   week      99999 non-null  int64 
 5   month     99999 non-null  int64 
 6   quarter   99999 non-null  int64 
 7   year      99999 non-null  int64 
dtypes: int64(6), object(2)
memory usage: 6.1+ MB


In [85]:
df_time['date'] = pd.to_datetime(df_time['date'])
df_time['date']

0       2017-05-20 14:56:00
1       2015-01-30 22:14:00
2       2020-03-14 02:34:00
3       2018-04-27 12:19:00
4       2018-04-14 10:43:00
                ...        
99994   2015-01-04 17:22:00
99995   2015-01-17 21:09:00
99996   2020-01-23 14:42:00
99997   2015-10-31 13:52:00
99998   2017-06-08 08:59:00
Name: date, Length: 99999, dtype: datetime64[ns]

In [86]:
df_time['time'] = df_time['date'].dt.time
df_time['date'] = df_time['date'].dt.date
df_time.head()

time_key        date  hour  day  week  month  quarter  year      time
0   T00001  2017-05-20    14   20     3      5        2  2017  14:56:00
1   T00002  2015-01-30    22   30     4      1        1  2015  22:14:00
2   T00003  2020-03-14     2   14     2      3        1  2020  02:34:00
3   T00004  2018-04-27    12   27     4      4        2  2018  12:19:00
4   T00005  2018-04-14    10   14     2      4        2  2018  10:43:00

In [87]:
df_time['week'] = df_time['week'].replace({
    "3rd Week": 3,
    "4th Week": 4,
    "2nd Week": 2,
    "1st Week": 1
})
df_time['week'] = df_time['week'].astype(int)

In [88]:
df_time['quarter'] = df_time['quarter'].replace({
    "Q1": 1,
    "Q2": 2,
    "Q3": 3,
    "Q4": 4})

df_time['quarter'] = df_time['quarter'].astype(int)

In [89]:
df_time['date'] = pd.to_datetime(df_time['date'])
df_time['quarter'] = df_time['quarter'].astype('category') 
df_time['time'] = pd.to_datetime(df_time['time'], format='%H:%M:%S').dt.time

In [90]:
df_time.dtypes

time_key            object
date        datetime64[ns]
hour                 int64
day                  int64
week                 int32
month                int64
quarter           category
year                 int64
time                object
dtype: object

In [91]:
df_time.sample(5)

time_key       date  hour  day  week  month quarter  year      time
12170  T012171 2015-02-24     8   24     4      2       1  2015  08:34:00
57782  T057783 2015-10-30     3   30     4     10       4  2015  03:03:00
75067  T075068 2017-07-26    19   26     4      7       3  2017  19:20:00
15588  T015589 2019-10-21     6   21     3     10       4  2019  06:10:00
97164  T097165 2020-04-30     3   30     4      4       2  2020  03:54:00

## **Summary**

| Category             | State | Action |
|----------------------|-------|--------|
| Number of duplicates | None  | None   |
| Null values          | None  | None   |
| Date column data type| `str`   | Changed to the propper datetime dtype |
| Date column          | `datetime` | Changed to only date |
| Time column          | Doesn't exist | Created from date column |
| Week number          | `str`   | Changed to `int` |
| Quarter number       | `str`   | Changed to `int` |

-----------------------------------------

# **Transaction**

In [92]:
df_trans.head()

payment_key trans_type                         bank_name
0        P001       cash                               NaN
1        P002       card                   AB Bank Limited
2        P003       card  Bangladesh Commerce Bank Limited
3        P004       card                 Bank Asia Limited
4        P005       card                 BRAC Bank Limited

In [93]:
df_trans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39 entries, 0 to 38
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   payment_key  39 non-null     object
 1   trans_type   39 non-null     object
 2   bank_name    38 non-null     object
dtypes: object(3)
memory usage: 1.0+ KB


In [94]:
df_trans['bank_name'] = df_trans['bank_name'].fillna('Unknown')

## **Summary**

| Category             | State | Action |
|----------------------|-------|--------|
| Number of duplicates | None  | None   |
| Null values          | 1 (bank name)  | Filled with `Unknown`   |

------------------------------------------------

## **Clean Data to be added to SQL database**

In [95]:
df_time.to_csv(r"C:\Mine\My Projects\E-commerce database\CSV\time.csv", index=False)
df_fact.to_csv(r"C:\Mine\My Projects\E-commerce database\CSV\fact.csv", index=False)
df_item.to_csv(r"C:\Mine\My Projects\E-commerce database\CSV\item.csv", index=False)
df_store.to_csv(r"C:\Mine\My Projects\E-commerce database\CSV\store.csv", index=False)
df_customer.to_csv(r"C:\Mine\My Projects\E-commerce database\CSV\customer.csv", index=False)
df_trans.to_csv(r"C:\Mine\My Projects\E-commerce database\CSV\trans.csv", index=False)